research/object_detection/meta_architectures/context_rcnn_lib_v2.py

```diff
@@ -0,0 +1,217 @@
+# Lint as: python3
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Library functions for ContextRCNN."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow.compat.v1 as tf
+
+# The negative value used in padding the invalid weights.
+_NEGATIVE_PADDING_VALUE = -100000
+
+KEY_NAME = 'key'
+VALUE_NAME = 'val'
+QUERY_NAME = 'query'
+FEATURE_NAME = 'feature'
+
+class ContextProjection(tf.keras.layers.Layer):
+  """Custom layer to do batch normalization and projection."""
+  def __init__(self, projection_dimension, freeze_batchnorm, **kwargs):
+    self.batch_norm = tf.keras.layers.BatchNormalization(
+        epsilon=0.001,
+        center=True,
+        scale=True,
+        momentum=0.97,
+        trainable=(not freeze_batchnorm))
+    self.projection = tf.keras.layers.Dense(units=projection_dimension,
+                                            activation=tf.nn.relu6,
+                                            use_bias=True)
+    super(ContextProjection, self).__init__(**kwargs)
+
+  def build(self, input_shape):
+    self.batch_norm.build(input_shape)
+    self.projection.build(input_shape)
+
+  def call(self, input_features, is_training=False):
+    return self.projection(self.batch_norm(input_features, is_training))
+
+class AttentionBlock(tf.keras.layers.Layer):
+  """Custom layer to perform all attention."""
+  def __init__(self, bottleneck_dimension, attention_temperature,
+               freeze_batchnorm, output_dimension=None, **kwargs):
+    self.key_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self.val_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self.query_proj = ContextProjection(bottleneck_dimension, freeze_batchnorm)
+    self.feature_proj = None
+    self.attention_temperature = attention_temperature
+    self.freeze_batchnorm = freeze_batchnorm
+    self.bottleneck_dimension = bottleneck_dimension
+    self.output_dimension = output_dimension
+    super(AttentionBlock, self).__init__(**kwargs)
+
+  def set_output_dimension(self, output_dim):
+    self.output_dimension = output_dim
+
+  def build(self, input_shapes):
+    pass
+
+  def call(self, input_features, is_training, valid_context_size):
```

Comment:
separate out input_features and context features. This will make the call site uniform and we don't have to check for TF1 or TF2

Reply:
+1